packages/core-api/src/extensions/traversal.ts

```diff
@@ -0,0 +1,139 @@
+/*
+ * Copyright 2020 Spotify AB
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import { isValidElement, ReactNode, ReactElement, Children } from 'react';
+
+export type Discoverer = (element: ReactElement) => ReactNode;
+
+export type Collector<Result, Context> = () => {
+  accumulator: Result;
+  visit(
+    accumulator: Result,
+    element: ReactElement,
+    parent: ReactElement,
+    context: Context,
+  ): Context;
+};
+
+/**
+ * A function that allows you to traverse a tree of React elements using
+ * varying methods to discover child nodes and collect data along the way.
+ */
+export function traverseElementTree<Results>(options: {
+  root: ReactElement;
+  discoverers: Discoverer[];
+  collectors: { [name in keyof Results]: Collector<Results[name], any> };
+}): Results {
+  const visited = new Set();
+  // const collectors = Object.fromEntries(Object.entries(options.collectors).map(([name, factory]) => [name, factory()]))
+  const collectors: {
+    [name in string]: ReturnType<Collector<any, any>>;
```

Comment:
`in string` 🤔  `: string`

Reply:
Not sure if the syntax is completely derived from this, but it's a bit of an explanation: https://www.typescriptlang.org/docs/handbook/advanced-types.html#mapped-types With mapped types and the `in` operator, the mindset would be that you're iterating through a union type, while `:` is that it's of the type `string`.

In this case it's really just used for consistency and laziness though, the correct type would really be `name in keyof Results`, but that'd force a cast of the RHS as it doesn't have all the keys present yet. Didn't consider it to be worth the noise

Reply:
Yeah so, it works because TS realizes it matches due to string widening the keyof T type... It's like it does an upcast of sorts